File: challenges/week2.01-matrix/ikasty.scala
// Runtime: 7052 ms, faster than 8.33% of Scala online submissions
// Memory Usage: 62.3 MB, less than 50.00% of Scala online submissions
import scala.collection.mutable.Queue

object Solution {
    def updateMatrix(matrix: Array[Array[Int]]): Array[Array[Int]] = {
        val result: Array[Array[Int]] = Array.fill(matrix.size, matrix.head.size)(-1)
        val queue = Queue[(Int, Int)]()

        // matrix에 0이 들어간 모든 좌표를 구한다
        val init =
            for (i <- (0 until matrix.size); j <- (0 until matrix.head.size) if matrix(i)(j) == 0)
                yield (i, j)

        // 해당하는 좌표의 result에 전부 0을 넣고, queue에도 넣는다
        init.foreach{ case (i, j) => result(i)(j) = 0 }
        queue ++= init

        while (!queue.isEmpty) {
            val (x, y) = queue.dequeue
            // 후보 좌표 중에서...
            val candidate = List((x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1))
                // matrix 범위 내에 있으며
               .filter(_._1 >= 0)
               .filter(_._2 >= 0)
               .filter(_._1 < matrix.size)
               .filter(_._2 < matrix.head.size)
                // 방문하지 않은 노드이며
               .filter{ case (i, j) => result(i)(j) < 0 }
                // 큐에 넣지 않은 노드라면
               .filter(!queue.contains(_))

            // 현재 result에서 1 증가시키고 큐에 넣기
            candidate.foreach{ case (i, j) => result(i)(j) = result(x)(y) + 1 }
            queue ++= candidate
        }

        // 리턴
        result
    }
}

File: challenges/week5.sum-of-absolute-differences/ikasty.py
# Runtime: 912 ms, faster than 92.39% of Python3 online submissions
# Memory Usage: 29.4 MB, less than 79.31% of Python3 online submissions
from itertools import accumulate

class Solution:
    def getSumAbsoluteDifferences(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [0] * n
        result[0] = sum(nums) - nums[0] * n

        left_sum, right_sum = nums[0], sum(nums) - nums[0]
        for i in range(1, n):
            right_sum -= nums[i]
            # result  = right_sum - left_sum - nums[i] * (len(right) - len(left))
            result[i] = right_sum - left_sum - nums[i] * ((n-i - 1) - i)
            left_sum += nums[i]

        return result